inference-engine/src/gna_plugin/memory/gna_memory_state.cpp

```diff
@@ -0,0 +1,112 @@
+            }
+        }
+
+        return state_precision;
+    }
+
+    void GNAMemoryState::SetState(InferenceEngine::Blob::Ptr newState) {
+        auto data_ptr = newState->cbuffer().as<void*>();
+        auto data_size = newState->byteSize();
+        auto data_elements = data_size / newState->element_size();
+        if (ALIGN64(state->reserved_size) != ALIGN64((data_size / (newState->element_size() / state->elementSizeBytes())))) {
+            THROW_GNA_EXCEPTION << "Failed to SetState. Sizes of new and old states do not match";
+        }
+
+        InferenceEngine::Precision state_precision = getPrecision();
+        auto new_state_precision = newState->getTensorDesc().getPrecision();
+
+        if (new_state_precision == state_precision) {
+            std::memcpy(state->gna_ptr, data_ptr, data_size);
+            return;
+        }
+
+        switch (state_precision) {
+        case InferenceEngine::Precision::I16: {
+            if (new_state_precision == InferenceEngine::Precision::FP32) {
+                auto quantized = InferenceEngine::getInjectedData<QuantizedLayerParams>(state->getInput());
+                auto scale_factor = quantized != nullptr ? quantized->_dst_quant.scale : 1.0f;
+                GNAPluginNS::ConvertToInt16(static_cast<int16_t*>(state->gna_ptr),
+                    newState->buffer().as<float*>(),
+                    1,
+                    data_elements,
+                    scale_factor);
+            } else {
+                THROW_GNA_EXCEPTION << "Failed to SetState. If old state precision is I16 only I16 and FP32 are allowed as new state precisions.";
+            }
+            break;
+        }
+        default:
+            THROW_GNA_EXCEPTION << "Failed to SetState. Incorrect new/old precision pair";
+        }
+    }
+
+    InferenceEngine::Blob::CPtr GNAMemoryState::GetLastState() const {
+        auto elements = state->reserved_size / state->elementSizeBytes();
+        InferenceEngine::Precision state_precision = getPrecision();
+
+        if (state->getInput() && state_precision == InferenceEngine::Precision::I16) {
+            auto quantized = InferenceEngine::getInjectedData<QuantizedLayerParams>(state->getInput());
+            auto scale_factor = quantized != nullptr ? quantized->_dst_quant.scale : 1.0f;
+
+            auto result_blob = make_blob_with_precision(InferenceEngine::TensorDesc(InferenceEngine::Precision::FP32,
+                InferenceEngine::SizeVector({ 1, elements }),
+                InferenceEngine::NC));
+
+            result_blob->allocate();
+            auto buffer = result_blob->buffer().as<float*>();
+            auto new_gna_ptr = static_cast<int16_t*>(state->gna_ptr);
+
+            for (int i = 0; i < elements; i++) {
+                buffer[i] = new_gna_ptr[i] / scale_factor;
+            }
+
+            return result_blob;
+        } else {
+            auto result_blob = make_blob_with_precision(InferenceEngine::TensorDesc(state_precision,
```

Comment:
and cache that blob - i would use cache per precision - but since memory state has single precision that not needed.

Reply:
caching of the blob might have consequences, if user keeps these blobs for reference somewhere, if we reuse the blob then values will change on user side without his knowledge, right now this method do not have any side effects, are you sure that you want introduce side effects just to avoid allocation? It will make it harder for the user. Basically getters should not change anything, while reusing blob will break this rule.

Reply:
this is actually already broken a bit, we probably should make a blob without gna_ptr, allocate the memory and copy, to make sure that state does not change on user side unknowingly

Reply:
Even worse, at the moment in case of FP32 user could use GetLastState() to unknowingly modify the state (or knowingly but in this case he should use SetState instead)

Reply:
returning blob by sharedptr and having name like get instead of clone / create / always implies that you are dealing with shared data - we can write that in contract. And user especially would like to manipulate with that data without calling set blob. Otherwise this is performance consern.  
GetLastState() only returns some representation - like custom precision blob with integers and scale factor. For having that we need to extra pad memory blob in front or end to keep this scale factor as one of element in blob occupying 4 bytes.
Usually user might clear scale factor but we will not use in gna_plugin.

Reply:
ok caching is possible, but looks required arch explanation, i'll create JIRA for that.